### ISSUE TITLE
REST endpoints don't insert torrents into MegaCache

### ISSUE THREAD

**Issue by ichorid** [Tribler/tribler#3985]
And it seems like it never worked as intended...

When the user adds a torrent from a torrent file or a magnet link, the torrent should be cached in `torrent_db`. **In 95% cases, it is not**.
Basically, DB entries are not added to the database *even on a forced DB commit*, but sometimes (in that 5% of cases) they do. We never noticed this until we disabled Dispersy, because Dispersy does _a lot_ of DB i/o, and it seems that eventually, it's entries are getting through, so the database contents seem legit.

Steps to reproduce:
Method 1. Add this test to `test_sqlitecachedbhandler_torrents.py` and run it.
```python
    def test_something(self):
        s_infohash = unhexlify('ed81da94d21ad1b305133f2726cdaec5a57fed98')
        self.tdb.updateTorrent(s_infohash, relevance=3.1415926, category=u'Videoclips',
                               status=u'good', seeder=123, leecher=321,
                               last_tracker_check=1234567,
                               other_key1='abcd', other_key2=123)
        s_torrent = self.tdb.getTorrent(s_infohash)
        self.assertNotEqual(None, s_torrent)
```

Method 2. Run Tribler and then run this shell script:
```bash
#!/bin/bash
curl -X GET "http://localhost:8085/torrentinfo?uri=magnet%3A%3Fxt%3Durn%3Abtih%3A07f1555918ed051809f0075fedc0cd469a194c93%26tr%3Dhttp%253A%252F%252Facademictorrents.com%252Fannounce.php%26tr%3Dudp%253A%252F%252Ftracker.publicbt.com%253A80%252Fannounce%26tr%3Dudp%253A%252F%252Ftracker.openbittorrent.com%253A80%252Fannounce"
echo "\n"

curl -X GET "http://localhost:8085/torrents/07f1555918ed051809f0075fedc0cd469a194c93"
```
the output of the second command should not be "unknown torrent" if Megacache work as intended, because the first command should put that infohash into the DB.

Method 3. Get into Tribler GUI, enable experimental Chant channel, add some torrents to it and find your own channel through the search. Then start clicking on torrents and observe the "unknown torrent" error. Then try downloading the same torrent, and then go and click on the same torrent in the channel again. Observe "unknown torrent" error. There should be no error if the torrent was cached. However, there is a slight chance that the torrent _would be_ added to the database, and then the error on click would be gone.

Method 4. Disable Dispersy through `triblerd.conf` and delete `tribler.sdb`. Then launch Tribler as normal and add some torrents for downloading. Then exit Tribler, launch `sqlitebrowser` and look into `tribler.sdb`'s `torrents` table. Observe it cleand and devoid of any traces of cached torrents. Again, there is a slight chance that one of a few of the downoloaded torrents _would be_ added there.

**Comment by ichorid** [comment]
I've tested it up to v.6.5.2. It is still there.
Either I understand something wrong about how MegaCache was intended to work, or the thing was broken from the beginning...

**Comment by qstokkink** [comment]
> Either I understand something wrong about how MegaCache was intended to work, or the thing was broken from the beginning...

At least your first method should work, so the latter actually seems to be true; broken from the beginning.

**Comment by ichorid** [comment]
@xoriole , thanks for the explanation! But there is still this problem with how REST endpoints update MegaCache. I guess it was just never intended to be used without Dispersy...

**Comment by ichorid** [comment]
According to @egbertbouman, DB `commit` statements originally were everywhere in the Tribler code. But people were getting locking problems all the time. Then someone noticed that its mostly Dispersy that gets stuff into the database, and there was an idea to move all `commit`s to the Dispersy thread.
Now threads are gone, but `commit`s are still there. With time, all `commit`s in the other parts of the code died out as unnecessary. There is now even no `commit` statement at the shutdown of Tribler.
To be ready to remove Dispersy we have to add `commit` at Tribler shutdown and at periodic intervals.

**Comment by ichorid** [comment]
Mostly fixed by #3986 and #3989
